src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+    /// created by trying to do a pushdown on input 0, followed by input 1,
+    /// followed by input 2, and so on
+    next_new_join_column: usize,
+    /// An array to look up the column in the new join corresponding to
+    /// * a column from the old join. The lookup key is the column number
+    ///   in the old join
+    /// * a partial aggregation or a pushed down aggregation. The lookup
+    ///   key is the `old_total_columns` + `agg_num`
+    ///   The array gives correct answers as long as it is never used to
+    ///   query for a column that is not present in the new join
+    permutation: HashMap<usize, usize>,
+    /// Reduction pushdown may cause expression join keys or expression group
+    /// keys to become evaluated already in an input of the new join.
+    /// Maps expressions from the old join to where they are evaluated in the
+    /// new join.
+    evaluated_expressions: HashMap<MirScalarExpr, MirScalarExpr>,
+
+    // Aggregations to try to push down
+    /// For each input, the aggregations and filters that involve just that input.
+    /// Contains ((position of aggregation, the aggregation), filters)
+    single_input_aggs_to_push: Vec<(Vec<(usize, AggregateExpr)>, Vec<MirScalarExpr>)>,
+    /// Aggregations involving no inputs. They can be partially pushed down to
+    /// any input as long as the join key is not a primary key
+    /// Contains (position of aggregation, the aggregation)
+    no_input_aggs_to_push: Vec<(usize, AggregateExpr)>,
+
+    // Fields used to determine whether aggregations should not be pushed down
+    // into a particular input
+    /// Whether the join implementation is DeltaQuery. Avoid pushdowns that
+    /// would invalidate the DeltaQuery plan.
+    is_delta: bool,
+    /// The unique keys of each of the old inputs. Avoid pushdowns where there
+    /// exists a unique key that is a subset of the group keys
+    old_input_keys: Vec<Vec<Vec<usize>>>,
+
+    // Other helper fields
+    /// The equivalences of the old join, with all single-input filters removed
+    equivalences: Vec<Vec<MirScalarExpr>>,
+    /// The support of aggregations that we do not try to push down, sorted by
+    /// input. Inner reduces must not cause these columns to be removed from
+    /// the join.
+    aggs_not_pushed_support: Vec<HashSet<usize>>,
+    /// The mapping of global to local columns in the old join
+    old_join_mapper: JoinInputMapper,
+    /// For each input, stores the inputs it is bound to and the (global) join
+    /// keys that bind it to the particular input. This is a `BTreeMap` so that
+    /// the resulting plan is deterministic.
+    join_graph: Vec<BTreeMap<usize, Vec<MirScalarExpr>>>,
+    /// whether aggregations were pushed down to any input
+    pushdown_succeeded: bool,
+}
+
+impl<'a> ReductionPusher<'a> {
+    /// Creates a struct for pushing down reductions. Returns `None` if the
+    /// relation is ineligible for reduction pushdown.
+    fn try_new(
```

Comment:
Very minor, but I think Rust style uses `try_` for things that may return a `Result`. I could be wrong about that, so feel free to ignore!